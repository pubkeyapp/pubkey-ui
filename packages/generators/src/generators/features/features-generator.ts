import { formatFiles, Tree } from '@nx/devkit'
import { join } from 'path'
import featureGenerator from '../feature/feature-generator'
import { FeaturesGeneratorSchema } from './features-generator-schema'
import { featureTypes } from '../feature/feature-types'

export async function featuresGenerator(tree: Tree, options: FeaturesGeneratorSchema) {
  const prefix = options.prefix ?? 'ui'
  const exports: string[] = []
  for (const type of featureTypes) {
    const target = type
    await featureGenerator(tree, {
      directory: join(options.directory, target),
      prefix,
      type,
      uiImport: options.uiImport ?? '@pubkey-ui/core',
    })
    exports.push(`export * from './${target}/${type}-feature'`)
  }
  tree.write(join(options.directory, 'index.ts'), exports.join('\n') + '\n')
  await formatFiles(tree)
}

export default featuresGenerator
