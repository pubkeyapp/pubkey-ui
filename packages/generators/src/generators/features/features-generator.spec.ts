import { Tree } from '@nx/devkit'
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing'
import { getRecursiveFileContents } from '../../'
import { featuresGenerator } from './features-generator'
import { FeaturesGeneratorSchema } from './features-generator-schema'

describe('features generator', () => {
  let tree: Tree
  const options: FeaturesGeneratorSchema = { directory: 'test', prefix: 'ui' }

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace()
  })

  it('should run successfully', async () => {
    await featuresGenerator(tree, { ...options })

    const contents = getRecursiveFileContents(tree, '.')
    expect(contents).toMatchSnapshot()
  })
})
