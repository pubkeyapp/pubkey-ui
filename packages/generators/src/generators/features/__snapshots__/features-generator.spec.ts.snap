// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features generator should run successfully 1`] = `
{
  ".prettierrc": {
    "content": [
      "{ "singleQuote": true }",
    ],
    "isBinary": false,
    "path": "./.prettierrc",
  },
  "nx.json": {
    "content": [
      "{",
      ""affected": { "defaultBase": "main" },",
      ""targetDefaults": {",
      ""build": { "cache": true },",
      ""lint": { "cache": true },",
      ""e2e": { "cache": true }",
      "}",
      "}",
    ],
    "isBinary": false,
    "path": "./nx.json",
  },
  "package.json": {
    "content": [
      "{",
      ""name": "@proj/source",",
      ""dependencies": {},",
      ""devDependencies": {}",
      "}",
    ],
    "isBinary": false,
    "path": "./package.json",
  },
  "test": {
    "children": {
      "demo": {
        "children": {
          "demo-feature-alerts.tsx": {
            "content": [
              "import { SimpleGrid } from '@mantine/core';",
              "import {",
              "UiCard,",
              "UiAlert,",
              "UiError,",
              "UiInfo,",
              "UiSuccess,",
              "UiWarning,",
              "} from '@pubkey-ui/core';",
              "export function DemoFeatureAlerts() {",
              "return (",
              "<UiCard title="Alerts">",
              "<SimpleGrid cols={2}>",
              "<UiError title="Error" message="Hello World" />",
              "<UiInfo title="Info" message="Hello World" />",
              "<UiSuccess title="Success" message="Hello World" />",
              "<UiWarning title="Warning" message="Hello World" />",
              "<UiAlert title="Alert" message="Hello World" />",
              "</SimpleGrid>",
              "</UiCard>",
              ");",
              "}",
            ],
            "isBinary": false,
            "path": "./test/demo/demo-feature-alerts.tsx",
          },
          "demo-feature-card.tsx": {
            "content": [
              "import { UiCard, UiStack } from '@pubkey-ui/core';",
              "export function DemoFeatureCard() {",
              "return (",
              "<UiStack>",
              "<UiCard withBorder title="Card" shadow="lg">",
              "CARD CONTENT",
              "</UiCard>",
              "</UiStack>",
              ");",
              "}",
            ],
            "isBinary": false,
            "path": "./test/demo/demo-feature-card.tsx",
          },
          "demo-feature-copy.tsx": {
            "content": [
              "import { UiCard, UiCopy, UiStack } from '@pubkey-ui/core';",
              "export function DemoFeatureCopy() {",
              "return (",
              "<UiCard title="Copy">",
              "<UiStack>",
              "<UiCopy text="Hello PubKey UI" />",
              "</UiStack>",
              "</UiCard>",
              ");",
              "}",
            ],
            "isBinary": false,
            "path": "./test/demo/demo-feature-copy.tsx",
          },
          "demo-feature-debug.tsx": {
            "content": [
              "import { SimpleGrid } from '@mantine/core';",
              "import { UiCard, UiDebug, UiDebugModal } from '@pubkey-ui/core';",
              "export function DemoFeatureDebug() {",
              "return (",
              "<UiCard title="Debug">",
              "<SimpleGrid cols={2}>",
              "<UiDebug data={{ foo: 'bar' }} open hideButton />",
              "<UiDebug data={{ foo: 'bar' }} />",
              "<UiDebug data={{ foo: 'bar' }} open />",
              "<UiDebugModal data={{ foo: 'bar' }} />",
              "</SimpleGrid>",
              "</UiCard>",
              ");",
              "}",
            ],
            "isBinary": false,
            "path": "./test/demo/demo-feature-debug.tsx",
          },
          "demo-feature-group.tsx": {
            "content": [
              "import { Button, SimpleGrid } from '@mantine/core';",
              "import { UiCard, UiGroup } from '@pubkey-ui/core';",
              "export function DemoFeatureGroup() {",
              "return (",
              "<UiCard title="Group">",
              "<SimpleGrid cols={2}>",
              "<UiGroup justify="space-between">",
              "<Button>Save</Button> <Button variant="default">Cancel</Button>",
              "</UiGroup>",
              "<UiGroup gap="xs">",
              "<Button>Save</Button> <Button variant="default">Cancel</Button>",
              "</UiGroup>",
              "<UiGroup gap="sm">",
              "<Button>Save</Button> <Button variant="default">Cancel</Button>",
              "</UiGroup>",
              "<UiGroup gap="md">",
              "<Button>Save</Button> <Button variant="default">Cancel</Button>",
              "</UiGroup>",
              "<UiGroup gap="lg">",
              "<Button>Save</Button> <Button variant="default">Cancel</Button>",
              "</UiGroup>",
              "<UiGroup gap="xl">",
              "<Button>Save</Button> <Button variant="default">Cancel</Button>",
              "</UiGroup>",
              "</SimpleGrid>",
              "</UiCard>",
              ");",
              "}",
            ],
            "isBinary": false,
            "path": "./test/demo/demo-feature-group.tsx",
          },
          "demo-feature-logo.tsx": {
            "content": [
              "import { SimpleGrid } from '@mantine/core';",
              "import {",
              "UiCard,",
              "UiLogoType,",
              "UiLogoTypeBlack,",
              "UiLogoTypeWhite,",
              "} from '@pubkey-ui/core';",
              "export function DemoFeatureLogo() {",
              "return (",
              "<UiCard title="Logo">",
              "<SimpleGrid cols={2}>",
              "<UiLogoType height={64} />",
              "<UiLogoTypeBlack height={64} />",
              "<UiLogoTypeWhite height={64} />",
              "</SimpleGrid>",
              "</UiCard>",
              ");",
              "}",
            ],
            "isBinary": false,
            "path": "./test/demo/demo-feature-logo.tsx",
          },
          "demo-feature-search-input.tsx": {
            "content": [
              "import { SimpleGrid } from '@mantine/core';",
              "import { UiCard, UiGroup, UiSearchInput } from '@pubkey-ui/core';",
              "export function DemoFeatureSearchInput() {",
              "return (",
              "<UiCard title="SearchInput">",
              "<SimpleGrid cols={2}>",
              "<UiSearchInput />",
              "<UiSearchInput icon={{ radius: 0 }} text={{ radius: 0 }} />",
              "</SimpleGrid>",
              "<UiSearchInput />",
              "<UiGroup>",
              "<UiSearchInput />",
              "<UiSearchInput />",
              "<UiSearchInput />",
              "<UiSearchInput />",
              "</UiGroup>",
              "</UiCard>",
              ");",
              "}",
            ],
            "isBinary": false,
            "path": "./test/demo/demo-feature-search-input.tsx",
          },
          "demo-feature-stack.tsx": {
            "content": [
              "import { Button, SimpleGrid } from '@mantine/core';",
              "import { UiCard, UiStack } from '@pubkey-ui/core';",
              "export function DemoFeatureStack() {",
              "return (",
              "<UiCard title="Stack">",
              "<SimpleGrid cols={2}>",
              "<UiStack justify="space-between">",
              "<Button>Save</Button> <Button variant="default">Cancel</Button>",
              "</UiStack>",
              "<UiStack gap="xs">",
              "<Button>Save</Button> <Button variant="default">Cancel</Button>",
              "</UiStack>",
              "<UiStack gap="sm">",
              "<Button>Save</Button> <Button variant="default">Cancel</Button>",
              "</UiStack>",
              "<UiStack gap="md">",
              "<Button>Save</Button> <Button variant="default">Cancel</Button>",
              "</UiStack>",
              "<UiStack gap="lg">",
              "<Button>Save</Button> <Button variant="default">Cancel</Button>",
              "</UiStack>",
              "<UiStack gap="xl">",
              "<Button>Save</Button> <Button variant="default">Cancel</Button>",
              "</UiStack>",
              "</SimpleGrid>",
              "</UiCard>",
              ");",
              "}",
            ],
            "isBinary": false,
            "path": "./test/demo/demo-feature-stack.tsx",
          },
          "demo-feature-tab-routes.tsx": {
            "content": [
              "import { SimpleGrid } from '@mantine/core';",
              "import { UiCard, UiTabRoutes } from '@pubkey-ui/core';",
              "export function DemoFeatureTabRoutes() {",
              "return (",
              "<UiCard title="Time">",
              "<UiTabRoutes",
              "baseUrl="/demo"",
              "tabs={[",
              "{",
              "value: 'overview',",
              "label: 'Overview',",
              "component: (",
              "<SimpleGrid cols={2} spacing="md">",
              "<UiCard title="Overview">Overview</UiCard>",
              "</SimpleGrid>",
              "),",
              "},",
              "{",
              "value: 'content',",
              "label: 'Content',",
              "component: (",
              "<SimpleGrid cols={2} spacing="md">",
              "<UiCard title="Content">Content</UiCard>",
              "</SimpleGrid>",
              "),",
              "},",
              "{",
              "value: 'settings',",
              "label: 'Settings',",
              "component: (",
              "<SimpleGrid cols={2} spacing="md">",
              "<UiCard title="Settings">Settings</UiCard>",
              "</SimpleGrid>",
              "),",
              "},",
              "]}",
              "/>",
              "</UiCard>",
              ");",
              "}",
            ],
            "isBinary": false,
            "path": "./test/demo/demo-feature-tab-routes.tsx",
          },
          "demo-feature-time.tsx": {
            "content": [
              "import { SimpleGrid } from '@mantine/core';",
              "import { UiCard, UiTime } from '@pubkey-ui/core';",
              "export function DemoFeatureTime() {",
              "return (",
              "<UiCard title="Time">",
              "<SimpleGrid cols={2}>",
              "<UiTime",
              "date={new Date(new Date().setMonth(new Date().getMonth() - 1))}",
              "/>",
              "<UiTime date={new Date(new Date().setDate(new Date().getDate() - 7))} />",
              "<UiTime date={new Date(new Date().setDate(new Date().getDate() - 2))} />",
              "<UiTime date={new Date(new Date().setDate(new Date().getDate() - 1))} />",
              "<UiTime date={new Date()} />",
              "<UiTime date={new Date(new Date().setDate(new Date().getDate() + 1))} />",
              "<UiTime date={new Date(new Date().setDate(new Date().getDate() + 7))} />",
              "<UiTime",
              "date={new Date(new Date().setMonth(new Date().getMonth() + 1))}",
              "/>",
              "<UiTime",
              "date={new Date(new Date().setFullYear(new Date().getFullYear() + 1))}",
              "/>",
              "<UiTime date={new Date('2021-01-01')} />",
              "</SimpleGrid>",
              "</UiCard>",
              ");",
              "}",
            ],
            "isBinary": false,
            "path": "./test/demo/demo-feature-time.tsx",
          },
          "demo-feature-toast.tsx": {
            "content": [
              "import { Button, SimpleGrid } from '@mantine/core';",
              "import {",
              "UiCard,",
              "toastError,",
              "toastInfo,",
              "toastSuccess,",
              "toastWarning,",
              "} from '@pubkey-ui/core';",
              "export function DemoFeatureToast() {",
              "return (",
              "<UiCard title="Toast">",
              "<SimpleGrid cols={2}>",
              "<Button",
              "color="green"",
              "onClick={() => toastSuccess('This is a success toast')}",
              ">",
              "Toast Success",
              "</Button>",
              "<Button",
              "color="red"",
              "onClick={() => toastError('This is an error toast')}",
              ">",
              "Toast Error",
              "</Button>",
              "<Button",
              "color="yellow"",
              "onClick={() => toastWarning('This is a warning toast')}",
              ">",
              "Toast Warning",
              "</Button>",
              "<Button color="cyan" onClick={() => toastInfo('This is an info toast')}>",
              "Toast Info",
              "</Button>",
              "</SimpleGrid>",
              "</UiCard>",
              ");",
              "}",
            ],
            "isBinary": false,
            "path": "./test/demo/demo-feature-toast.tsx",
          },
          "demo-feature.tsx": {
            "content": [
              "import { UiContainer, UiStack } from '@pubkey-ui/core';",
              "import { DemoFeatureAlerts } from './demo-feature-alerts';",
              "import { DemoFeatureCard } from './demo-feature-card';",
              "import { DemoFeatureCopy } from './demo-feature-copy';",
              "import { DemoFeatureDebug } from './demo-feature-debug';",
              "import { DemoFeatureGroup } from './demo-feature-group';",
              "import { DemoFeatureLogo } from './demo-feature-logo';",
              "import { DemoFeatureSearchInput } from './demo-feature-search-input';",
              "import { DemoFeatureStack } from './demo-feature-stack';",
              "import { DemoFeatureTabRoutes } from './demo-feature-tab-routes';",
              "import { DemoFeatureTime } from './demo-feature-time';",
              "import { DemoFeatureToast } from './demo-feature-toast';",
              "export function DemoFeature() {",
              "return (",
              "<UiContainer>",
              "<UiStack>",
              "<DemoFeatureAlerts />",
              "<DemoFeatureCard />",
              "<DemoFeatureCopy />",
              "<DemoFeatureDebug />",
              "<DemoFeatureGroup />",
              "<DemoFeatureLogo />",
              "<DemoFeatureSearchInput />",
              "<DemoFeatureStack />",
              "<DemoFeatureTabRoutes />",
              "<DemoFeatureTime />",
              "<DemoFeatureToast />",
              "</UiStack>",
              "</UiContainer>",
              ");",
              "}",
            ],
            "isBinary": false,
            "path": "./test/demo/demo-feature.tsx",
          },
        },
        "path": "./test/demo",
      },
      "index.ts": {
        "content": [
          "export * from './demo/demo-feature';",
        ],
        "isBinary": false,
        "path": "./test/index.ts",
      },
    },
    "path": "./test",
  },
  "tsconfig.base.json": {
    "content": [
      "{",
      ""compilerOptions": {",
      ""paths": {}",
      "}",
      "}",
    ],
    "isBinary": false,
    "path": "./tsconfig.base.json",
  },
}
`;
